src/main/kotlin/calculator/Calculator.kt

```diff
@@ -1,58 +1,26 @@
 package calculator
 
-import calculator.exceptions.NumberSeparationException
 import calculator.exceptions.OperationBetweenNumbersException
 
 object Calculator {
     fun calculate(input: String?) = calculate(InputParser.validateInput(input))
 
     private fun calculate(inputList: List<String>): Double {
-        var currentCalculation = processNumber(inputList.first())
-        var pendingOperation: Operation? = null
+        val numbersList = processNumbers(inputList)
+        val operations: ArrayDeque<Operation> = ArrayDeque(processOperations(inputList))
 
-        for (index in 1..inputList.lastIndex) {
-            if (index % 2 == 0) {
-                // even element - number
-                val number = processNumber(inputList[index])
-                currentCalculation =
-                    performOperation(
-                        operation = requireNotNull(pendingOperation) { "Operation is required on this step" },
-                        a = currentCalculation,
-                        b = number,
-                    )
-            } else {
-                // odd - operation
-                pendingOperation = processOperation(inputList[index])
-            }
+        return numbersList.drop(1).fold(numbersList.first()) { result, number ->
+            operations.removeFirst().apply(a = result, b = number)
         }
-
-        return currentCalculation
     }
 
-    private fun performOperation(
-        operation: Operation,
-        a: Double,
-        b: Double,
-    ): Double =
-        when (operation) {
-            Operation.ADD -> a + b
-            Operation.SUBTRACT -> a - b
-            Operation.MULTIPLY -> a * b
-            Operation.DIVIDE -> a / b
-        }
-
-    private fun processNumber(inputNumber: String): Double = inputNumber.toDoubleOrNull() ?: throw OperationBetweenNumbersException()
+    private fun formatNumber(inputNumber: String): Double = inputNumber.toDoubleOrNull() ?: throw OperationBetweenNumbersException()
```

Comment:
It might be easier for other developers to read this function if it were placed below `processNumbers()`.